nmigen/build/generic_platform.py

```diff
@@ -0,0 +1,161 @@
+from collections import defaultdict, OrderedDict
+
+from .. import *
+from ..hdl.rec import *
+from ..lib.io import *
+
+from .dsl import *
+from .tools import *
+
+
+__all__ = ["ConstraintError", "ConstraintManager"]
+
+
+class ConstraintError(Exception):
+    pass
+
+
+class ConstraintManager:
+    def __init__(self, resources):
+        self.available    = defaultdict(dict)
+        self.matched      = OrderedDict()
+        self.clocks       = OrderedDict()
+        self.tristates    = []
+        self.diffpairs    = []
+        self.ports        = []
+        self.inouts       = []
+
+        self.add_resources(resources)
+
+    def add_resources(self, resources):
+        for r in resources:
+            if not isinstance(r, Resource):
+                raise TypeError("Object {!r} is not a Resource".format(r))
+            if r.number in self.available[r.name]:
+                raise NameError("Trying to add {!r}, but {!r} has the same name and number"
+                                .format(r, self.available[r.name][r.number]))
+            self.available[r.name][r.number] = r
+
+    def add_period(self, resource, period):
+        if not isinstance(resource, Resource):
+            raise TypeError("Object {!r} is not a Resource".format(resource))
+        if isinstance(resource.io[0], Subsignal):
+            raise ConstraintError("Cannot constrain Resource ('{}', {}) to a period of {} ns "
```

Comment:
It would be more clear to say "Cannot constrain period of resource … because it has subsignals", since there is no period that would be valid here at all.